### DISCUSSION TITLE
Is there any methods in 2.x can be used to implement a function like useViewportState in 3.x?

### DISCUSSION THREAD

**Discussion by Songhao-LI** [annotorious/annotorious#452] [Q&A] [answered]
https://annotorious.dev/react/hooks-reference/
![image](https://github.com/user-attachments/assets/238895d5-dee0-482c-be97-6dc74b4e0c7c)

I will be very grateful if anyone can tell me how to do!!

**Comment by rsimon** [comment, marked as answer]
It's certainly possible. But not trivial to implement. You'd need to listen to the `update-viewport` event on the OpenSeadragon viewer, then determine the current viewport bounds (using OpenSeadragon's API) in the coordinate space of the base image, and finally filter all annotations based on whether they intersect the bounds. 

If you are not expecting to have many annotations, then simply looping through all annotations will work. (If you expect to have many, you'd probably have to build a 2D index and keep it in sync as annotations are added, updated and removed. Annotorious v3 does that. But it's probably not necessary unless you go over several 100 if not 1000 annotations.)

**Reply by Songhao-LI** [reply]
Thanks!!
